### ISSUE TITLE
shellwords.pl cannot handle trailing backslash

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#8683]
Migrated from [rt.perl.org#40921](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=40921) (status was 'resolved')

Searchable as RT40921$

**Comment by p5pRT** [comment]
### From alexander_bluhm@genua.de

#### Created by bluhm@genua.de



The library function shellwords does not terminate\, if the given
string ends with a backslash\.

Reproduce&#8203;:
perl \-e 'require "shellwords\.pl"; shellwords\("foo\\\\"\)'

Fix&#8203;:
<details><summary>Inline Patch</summary>

```diff
--- /usr/libdata/perl5/shellwords.pl	Wed Mar 15 08:31:11 2006
+++ shellwords.pl	Fri Nov 17 18:02:55 2006
@@ -30,7 +30,7 @@
 	    elsif (/^'/) {
 		die "Unmatched single quote: $_\n";
 	    }
-	    elsif (s/^\\(.)//) {
+	    elsif (s/^\\(.?)//) {
 		$snippet = $1;
 	    }
 	    elsif (s/^([^\s\\'"]+)//) {
```
</details>


<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=low

Site configuration information for perl v5.8.6:

Configured by root at Thu Jan  1  0:00:00 UTC 1970.

Summary of my perl5 (revision 5 version 8 subversion 6) configuration:
  Platform:
    osname=openbsd, osvers=3.9, archname=i386-openbsd
    uname='openbsd'
    config_args='-dsE -Dopenbsd_distribution=defined'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -fno-delete-null-pointer-checks -pipe -I/usr/local/include',
    optimize='-O2',
    cppflags='-fno-strict-aliasing -fno-delete-null-pointer-checks -pipe -I/usr/local/include'
    ccversion='', gccversion='3.3.5 (propolice)', gccosandvers='openbsd3.9'
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags ='-Wl,-E '
    libpth=/usr/lib
    libs=-lm -lutil -lc
    perllibs=-lm -lutil -lc
    libc=/usr/lib/libc.so.39.0, so=so, useshrplib=true, libperl=libperl.so.10.0
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-R/usr/libdata/perl5/i386-openbsd/5.8.6/CORE'
    cccdlflags='-DPIC -fPIC ', lddlflags='-shared -fPIC '

Locally applied patches:
    SUIDPERLIO1 - fix PERLIO_DEBUG buffer overflow (CAN-2005-0156)
    SPRINTF0 - fixes for sprintf formatting issues - CVE-2005-3962


@INC for perl v5.8.6:
    /usr/libdata/perl5/i386-openbsd/5.8.6
    /usr/local/libdata/perl5/i386-openbsd/5.8.6
    /usr/libdata/perl5
    /usr/local/libdata/perl5
    /usr/local/libdata/perl5/site_perl/i386-openbsd
    /usr/libdata/perl5/site_perl/i386-openbsd
    /usr/local/libdata/perl5/site_perl
    /usr/libdata/perl5/site_perl
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.8.6:
    HOME=/home/bluhm
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/bin:/bin:/usr/sbin:/sbin:/usr/X11R6/bin:/usr/local/bin:/home/bluhm/bin
    PERL_BADLANG (unset)
    SHELL=/bin/ksh

```
</details>

**Comment by p5pRT** [comment]
### From @rgs

shellwords\.pl is now deprecated and only a thin wrapper around the
module Text&#8203;::ParseWords\.

**Comment by p5pRT** [comment]
### From @rgarcia

shellwords\.pl is now deprecated and only a thin wrapper around the
module Text&#8203;::ParseWords\.

**Comment by p5pRT** [comment]
@rgs - Status changed from 'new' to 'resolved'

**Comment by p5pRT** [comment]
### From @nwc10

On Sat\, Nov 18\, 2006 at 05&#8203;:04&#8203;:37AM \-0800\, Rafael Garcia\-Suarez via RT wrote&#8203;:
> shellwords\.pl is now deprecated and only a thin wrapper around the
> module Text&#8203;::ParseWords\.

Which\, it seems\, contains the code from shellwords\.pl

On Fri\, Nov 17\, 2006 at 09&#8203;:31&#8203;:48AM \-0800\, alexander\_bluhm @&#8203; genua\. de wrote&#8203;:

> Fix&#8203;:
> \-\-\- /usr/libdata/perl5/shellwords\.pl    Wed Mar 15 08&#8203;:31&#8203;:11 2006
> \+\+\+ shellwords\.pl    Fri Nov 17 18&#8203;:02&#8203;:55 2006
> @&#8203;@&#8203; \-30\,7 \+30\,7 @&#8203;@&#8203;
>          elsif \(/^'/\) \{
>          die "Unmatched single quote&#8203;: $\_\\n";
>          \}
> \-        elsif \(s/^\\\\\(\.\)//\) \{
> \+        elsif \(s/^\\\\\(\.?\)//\) \{
>          $snippet = $1;
>          \}
>          elsif \(s/^\(\[^\\s\\\\'"\]\+\)//\) \{

I applied this to Text&#8203;::ParseWords with change 29316\.

Thanks\.

Nicholas Clark
